src/librustc_mir_build/build/expr/as_operand.rs

```diff
@@ -75,39 +75,49 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
     where
         M: Mirror<'tcx, Output = Expr<'tcx>>,
     {
-        let tcx = self.hir.tcx();
+        let this = self;
+        let expr = this.hir.mirror(expr);
 
-        let local_scope = self.local_scope();
-        let expr = self.hir.mirror(expr);
+        if let ExprKind::Scope { region_scope, lint_level, value } = expr.kind {
+            let source_info = this.source_info(expr.span);
+            let region_scope = (region_scope, source_info);
+            return this
+                .in_scope(region_scope, lint_level, |this| this.as_call_operand(block, value));
+        }
+
+        let tcx = this.hir.tcx();
+        let local_scope = this.local_scope();
```

Comment:
So the bug is about this line -- this gets the scope in which the returned value should be valid. The intent is to read the "enclosing scope at the time that `as_call_operand` was called. But because we are now recursing into scopes, what we're getting is an inner scope instead -- i.e., something specific to this particular operand.

Basically for a call, there are (at least) two scopes involved:

```
foo(    // <---------------+ scope for `foo(*x)`
  *x,   // scope for `*x`  |
) // <---------------------+
```

The `local_scope` as we initially call `as_call_operand` is the outer scope, the one for `foo(*x)`, but once we recurse, `local_scope` refers to `*x`. 

I'll push the commit I used to test this theory, which also fixes the problem. It basically just passes in `local_scope` as a parameter. Maybe it'd be better even to rename it to `call_scope` or something like that.

Reply:
@nikomatsakis right, I've followed a setup similar to what we have in `as_operand`. Check it out and please send me your patch again because I've force pushed and if I'm not wrong you also wrote some comments and some other useful stuff :).